pkg/upgrade/upgrade_state.go

```diff
@@ -0,0 +1,279 @@
+	ctx context.Context, currentClusterState *ClusterUpgradeState, nodeStateName string) error {
+	m.Log.V(consts.LogLevelInfo).Info("ProcessReadyOrUnknownNodes")
+
+	for _, nodeState := range currentClusterState.NodeStates[nodeStateName] {
+		if GetPodGeneration(nodeState.DriverPod, m.Log) != nodeState.DriverDaemonSet.GetGeneration() {
+			err := ChangeNodeUpgradeState(ctx, nodeState.Node, UpgradeStateUpgradeRequired, m.K8sClient, m.Log)
+			if err != nil {
+				return err
+			}
+		} else if nodeStateName == UpgradeStateUnknown {
+			err := ChangeNodeUpgradeState(ctx, nodeState.Node, UpgradeStateReady, m.K8sClient, m.Log)
+			if err != nil {
+				return err
+			}
+		}
+		m.Log.V(consts.LogLevelInfo).Info("Node in ready state, upgrade not required",
+			"node", nodeState.Node.Name)
+	}
+	return nil
+}
+
+func (m *ClusterUpgradeStateManager) ProcessUpgradeRequiredNodes(
+	ctx context.Context, currentClusterState *ClusterUpgradeState, limit int) error {
+	m.Log.V(consts.LogLevelInfo).Info("ProcessUpgradeRequiredNodes")
+	for _, nodeState := range currentClusterState.NodeStates[UpgradeStateUpgradeRequired] {
+		if limit <= 0 {
+			m.Log.V(consts.LogLevelInfo).Info("Limit for new upgrades is exceeded, skipping the iteration")
+			break
+		}
+
+		err := ChangeNodeUpgradeState(ctx, nodeState.Node, UpgradeStateDrain, m.K8sClient, m.Log)
+		if err == nil {
+			limit--
+			m.Log.V(consts.LogLevelInfo).Info("Node waiting for drain",
+				"node", nodeState.Node.Name)
+		} else {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *ClusterUpgradeStateManager) ProcessDrainNodes(
+	ctx context.Context, currentClusterState *ClusterUpgradeState, drainSpec *v1alpha1.DrainSpec) error {
+	m.Log.V(consts.LogLevelInfo).Info("ProcessDrainNodes")
+	drainConfig := DrainConfiguration{
+		Spec:  drainSpec,
+		Nodes: make([]*v1.Node, 0, len(currentClusterState.NodeStates[UpgradeStateDrain])),
+	}
+	for _, nodeState := range currentClusterState.NodeStates[UpgradeStateDrain] {
+		drainConfig.Nodes = append(drainConfig.Nodes, nodeState.Node)
+	}
+
+	return m.DrainManager.ScheduleNodesDrain(ctx, &drainConfig)
+}
+
+func (m *ClusterUpgradeStateManager) ProcessPodRestartNodes(
+	ctx context.Context, currentClusterState *ClusterUpgradeState) error {
+	m.Log.V(consts.LogLevelInfo).Info("ProcessPodRestartNodes")
+
+	pods := make([]*v1.Pod, 0, len(currentClusterState.NodeStates[UpgradeStatePodRestart]))
+	for _, nodeState := range currentClusterState.NodeStates[UpgradeStatePodRestart] {
+		if GetPodGeneration(nodeState.DriverPod, m.Log) != nodeState.DriverDaemonSet.GetGeneration() {
+			// Pods should only be scheduled for restart if they are not terminating or restarting already
+			if nodeState.DriverPod.Status.Phase == "Running" {
+				pods = append(pods, nodeState.DriverPod)
+			}
+		} else if nodeState.DriverPod.Status.Phase == "Running" {
+			err := ChangeNodeUpgradeState(ctx, nodeState.Node, UpgradeStateReady, m.K8sClient, m.Log)
```

Comment:
maybe im missing something, but are we missing an upgrade state here (driverPodReadyUncordonNeeded or similar) ?

if a pod is restarted and now matches the generation of the daemonset, and it is now running. when will uncordon happen ? since here we change the upgrade state of the node to ready.

since in ProcessUncordon, we only uncordon pod restart and drain failed states.

Reply:
Added UncordonRequired state